Transformers: fix src and key padding mask bool regression

mikekgfb opened the pull request:
Summary: fix src and pad mask bool regression

This fixes a regression introduced previously with #92733. That PR unified testing of masks to remove Byte Tensors as permissible mask, introduced mask compatibility check, and mask conversion to FP mask.  The problem addressed in this PR was that after the first mask had been converted, a check for mask compatibility would fail.  

Test Plan: sandcastle & github

Differential Revision: D43782858

Fixes  https://github.com/pytorch/pytorch/issues/95702

drisspg commented on test/test_transformers.py:
```diff
@@ -2111,6 +2111,36 @@ def test_fused_kernels_nested_broadcasting_error_cases(self):
 
 
 
+    def test_padding_and_src_mask_bool(self):
```
can you move this test to the TestTransformers class

mikekgfb replied:
done

mikekgfb commented:
Fix issue 95702

malfet commented on test/test_transformers.py:
```diff
@@ -2111,6 +2143,7 @@ def test_fused_kernels_nested_broadcasting_error_cases(self):
 
 
 
+
```
Nit
```suggestion
```

malfet commented on torch/nn/functional.py:
```diff
@@ -5206,8 +5206,8 @@ def multi_head_attention_forward(
         mask=attn_mask,
         mask_name="attn_mask",
-        other_type=_none_or_dtype(key_padding_mask),
-        other_name="key_padding_mask",
```
Is there a reason to change `other_name`? (I guess it would affect serialization, wouldn't it?)

mikekgfb replied:
See description.  This is for cross checking types of masks.  After mask1 and mask2 have been cross checked, and mask1 has been converted, we would fail the check if we repeat the check at that point.

cpuhrsch commented on test/test_transformers.py:
```diff
@@ -510,6 +510,38 @@ def perm_fn(x):
+
+        assert(src_mask.dtype == padding_mask.dtype)
+
+        encoder(
```
What is this meant to test?

mikekgfb replied:
This is the tdefect converted to a test.  I guess the assert was supposed to prove that the dtypes of the twop masks are the same.  Can be removed in the unit test.

cpuhrsch commented on test/test_transformers.py:
```diff
@@ -510,6 +510,38 @@ def perm_fn(x):
+            torch.arange(3)[None, :].cpu() >= input_seq_len[:, None]
+        )
+
+        assert(src_mask.dtype == padding_mask.dtype)
```
```suggestion
        self.assertEqual(src_mask.dtype, padding_mask.dtype)
```

mikekgfb replied:
Removed, as per above.

cpuhrsch reviewed (approved):
See inline comments. cc @drisspg

malfet reviewed (approved):
It would be good to have a more detailed PR description, that mentions issue where it is reported(already mentioned), PR that caused it, and why passing `other_type` is wrong in this case.

cpuhrsch reviewed (changes requested):
Please address the comments before landing

drisspg commented on test/test_transformers.py:
```diff
@@ -510,6 +510,38 @@ def perm_fn(x):
+            torch.arange(3)[None, :].cpu() >= input_seq_len[:, None]
+        )
+
+        assert(src_mask.dtype == padding_mask.dtype)
```
This test passes currently. I think we should check that we no longer warning.

Commit:
fix src and pad mask bool regression (#96009)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/96009

fix src and pad mask bool regression

Test Plan: sandcastle & github

Reviewed By: cpuhrsch

Differential Revision: D43782858

fbshipit-source-id: 833e6d67cd2b194e5f030784c071536662a26d5e

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@pytorchbot merge

atalman commented:
cc @mikekgfb @ydwu4 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176
